File: java/problem-4/Problem4.java
import java.util.*;

public class Problem4 {
    public static void main(String[] args) {
        int answer = highestPalindromeProduct(100, 999);
        System.out.println(answer);
    }

    private static int highestPalindromeProduct(int lowestFactor, int highestFactor) {
        TreeSet<Integer> palindromes = new TreeSet<Integer>();

        for (int i = highestFactor; i >= lowestFactor; i--) {
            for (int j = highestFactor; j >= lowestFactor; j--) {
                int product = i * j;
                if (isPalindrome(product)) {
                    palindromes.add(product);
                }
            }
        }

        return palindromes.last();
    }

    private static boolean isPalindrome(int n) {
        String nString = String.valueOf(n);
        String nReversed = new StringBuilder(nString).reverse().toString();

        return (nString.equals(nReversed));
    }
}

File: go/problem-4/problem-4.go
package main

import (
    "fmt"
    "sort"
    "strconv"
)

func main() {
    palindromes := palindromeProducts(100, 999)
    sort.Ints(palindromes)
    answer := palindromes[len(palindromes) - 1]

    fmt.Println(answer)
}

func palindromeProducts(lowestFactor int, highestFactor int) []int {
    var palindromes []int

    for i := highestFactor; i >= lowestFactor; i-- {
        for j := highestFactor; j >= lowestFactor; j-- {
            product := i * j

            if (isPalindrome(product)) {
                palindromes = append(palindromes, product)
            }
        }
    }

    return palindromes
}

func isPalindrome(n int) bool {
    s := strconv.Itoa(n)

    // The index of the last character in the first half of the string.
    // For strings with an odd number of characters, this is the character
    // shared by both halves.
    firstHalfIndex := (len(s) + 1) / 2

    for i := 0; i < firstHalfIndex; i++ {
        character := s[i]
        characterOpposite := s[len(s) - 1 - i]
        if (character != characterOpposite) {
            return false
        }
    }

    return true
}
